File: mvvm/src/main/java/com/lee/mvvm/base/view/BaseFragment.java
package com.lee.mvvm.base.view;

import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProviders;
import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.os.Bundle;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.lee.mvvm.base.view.intf.IBaseView;
import com.lee.mvvm.utils.LifeCycleHelper;
import com.trello.rxlifecycle2.components.support.RxFragment;

public abstract class BaseFragment<B extends ViewDataBinding, VM extends ViewModel> extends RxFragment
        implements IBaseView {

    private View baseView;
    protected B binding;
    protected VM vm;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        if (baseView == null) {
            binding = DataBindingUtil.inflate(inflater, bindView(), container, false);
            baseView = binding.getRoot();
            initView();
            vm = initVM();
            loadData(getArguments());
            binding.setLifecycleOwner(this);
        } else if (container != null) {
            ViewGroup viewGroup = (ViewGroup) container.getParent();
            viewGroup.removeView(this.baseView);
        }
        return baseView;
    }

    @Override
    public LifeCycleHelper getDefaultLifeCycle() {
        return new LifeCycleHelper(this);
    }

    protected VM getDefaultVMInstance(Class<VM> vmClass) {
        if (getActivity() != null) {
            return ViewModelProviders.of(getActivity()).get(getActivity().getClass().getName(), vmClass);
        }
        return ViewModelProviders.of(this).get(getClass().getName(), vmClass);
    }

    /**
     * 获取到databinding将要绑定的布局Id
     *
     * @return layoutId
     */
    @LayoutRes
    protected abstract int bindView();

    /**
     * 进行可能的View初始化
     */
    protected abstract void initView();

    /**
     * 每一个View默认持有一个ViewModel
     *
     * @return VM: Nullable
     */
    protected abstract VM initVM();

    /**
     * 进行数据的初始化以及Databinding中的数据绑定
     *
     * @param dataFromIntent 从Intent中取到的数据
     */
    protected abstract void loadData(Bundle dataFromIntent);
}

File: mvvm/src/main/java/com/lee/mvvm/base/view/BaseFragment.kt
package com.lee.mvvm.base.view

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProviders
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import android.os.Bundle
import androidx.annotation.LayoutRes
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.lee.mvvm.base.view.intf.IBaseView
import com.lee.mvvm.utils.LifeCycleHelper
import com.trello.rxlifecycle3.components.support.RxFragment

abstract class BaseFragment<B : ViewDataBinding, VM : ViewModel> : RxFragment(), IBaseView {

    private var baseView: View? = null
    protected lateinit var binding: B
    protected lateinit var vm: VM

    override fun getDefaultLifeCycle() = LifeCycleHelper(this)

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        if (baseView == null) {
            binding = DataBindingUtil.inflate(inflater, bindView(), container, false)
            baseView = binding.root
            initView()
            vm = initVM()
            loadData(arguments)
            binding.setLifecycleOwner(this)
        } else if (container != null) {
            val viewGroup = container.parent as ViewGroup
            viewGroup.removeView(this.baseView)
        }
        return baseView
    }

    protected fun getDefaultVMInstance(vmClass: Class<VM>): VM {
        return if (activity != null) {
            ViewModelProviders.of(activity!!).get(activity!!.javaClass.name, vmClass)
        } else ViewModelProviders.of(this).get(javaClass.name, vmClass)
    }

    /**
     * 获取到databinding将要绑定的布局Id
     *
     * @return layoutId
     */
    @LayoutRes
    protected abstract fun bindView(): Int

    /**
     * 进行可能的View初始化
     */
    protected abstract fun initView()

    /**
     * 每一个View默认持有一个ViewModel
     *
     * @return VM: Nullable
     */
    protected abstract fun initVM(): VM

    /**
     * 进行数据的初始化以及Databinding中的数据绑定
     *
     * @param dataFromIntent 从Intent中取到的数据
     */
    protected abstract fun loadData(dataFromIntent: Bundle?)
}
